engine/server/engine/server/websocket_api_handler.go

```diff
@@ -245,32 +274,43 @@ func streamStarlarkLogsWithHTTP[T any](ctx echo.Context, streamerPool streaming.
 	}
 }
 
-func streamServiceLogsWithWebsocket(ctx echo.Context, streamer streaming.ServiceLogStreamer) {
-	websocket.Handler(func(ws *websocket.Conn) {
-		defer ws.Close()
-		err := streamer.Consume(func(logline *api_type.ServiceLogs) error {
-			err := websocket.JSON.Send(ws, logline)
-			if err != nil {
-				return err
-			}
-			return nil
+func streamServiceLogsWithWebsocket(ctx echo.Context, cors cors.Cors, streamer streaming.ServiceLogStreamer) {
+	wsPump, err := NewWebsocketPump[api_type.ServiceLogs](ctx, cors)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to start websocket connection")
+		ctx.Response().Header().Set(echo.HeaderContentType, echo.MIMEApplicationJSON)
+		enc := json.NewEncoder(ctx.Response())
+		ctx.Response().WriteHeader(http.StatusInternalServerError)
+		err = enc.Encode(api_type.ResponseInfo{
+			Type:    api_type.ERROR,
+			Message: "Failed to start websocket connection",
+			Code:    http.StatusInternalServerError,
 		})
-
 		if err != nil {
-			logrus.WithError(err).WithFields(logrus.Fields{
-				"stacktrace": fmt.Sprintf("%+v", err),
-				"services":   streamer.GetRequestedServiceUuids(),
-			}).Error("Failed to stream all data")
-			streamingErr := api_type.ResponseInfo{
-				Type:    api_type.ERROR,
-				Message: "Log streaming failed while sending the data",
-				Code:    http.StatusInternalServerError,
-			}
-			if err := websocket.JSON.Send(ws, streamingErr); err != nil {
-				logrus.WithError(err).Errorf("Failed to send value via websocket")
-			}
+			logrus.WithError(err).Error("Failed to encode error message")
 		}
-	}).ServeHTTP(ctx.Response(), ctx.Request())
+		return
+	}
+	defer wsPump.Close()
+	go wsPump.startPumping()
+
+	err = streamer.Consume(func(logline *api_type.ServiceLogs) error {
+		wsPump.pumpMessage(logline)
+		return nil
+	})
+
+	if err != nil {
+		logrus.WithError(err).WithFields(logrus.Fields{
```

Comment:
same for this error

Reply:
Here it's different, we are returning the error over the websocket channel. I can't use the same abstraction.